### PR TITLE
[#10917] Fix unexpected behaviour when reverting to template feedback path

### LINKED ISSUE #10917: Unexpected behaviour when reverting to template feedback path
**Environment**: Live server at V7.8.0

**Steps to reproduce**

1. Add new essay question to a feedback session
2. Change feedback path to custom feedback path
3. Attempt to revert feedback path to preset template provided

**Expected behaviour**
User interface reverts to template display prior to selecting custom feedback path

**Actual behaviour**
User interface does not change and the visibility options are reset to "deny all"

**Additional info**
Jan-16-2021 00-04-05

### PR DESCRIPTION
**Pull request by stapletonce**
Fixes #10917

**Outline of Solution**

In the changeGiverRecipientType on line 240 of /web/app/components/question-edit-form/question-edit-form.component.ts, I added an if statement that checks if a custom feedback path is currently being used the visibility settings will not change when it is reverted to its template settings.

https://user-images.githubusercontent.com/55817381/108639389-43eaa400-7462-11eb-939d-54c64c8f93f5.mov

### CODE REVIEW

**Review by Derek-Hardy — COMMENTED** [review]
@stapletonce One of the related e2e test has failed. It will be worthwhile to investigate whether to update e2e test workflow accordingly with latest change or to fix potential loophole in the current implementation.

It will be helpful to look into relevant parts of `InstructorFeedbackEditPageE2ETest.java` for debugging.

**Comment by stapletonce** [comment]
@Derek-Hardy Hello! I believe I have finished fixing this issue though I'm not sure if I have access to get the ongoing label off.

**Comment by stapletonce** [comment]
@madanalogy, @Derek-Hardy, @damithc This issue is no longer in progress and is ready for review (as far as I can tell!)

File: src/web/app/components/question-edit-form/question-edit-form.component.ts [hunk 1]
```diff
@@ -246,15 +246,25 @@ export class QuestionEditFormComponent implements OnInit {
...
-      recipientType: newRecipientType,
-      commonVisibilitySettingName: 'Please select a visibility option',
-      isUsingOtherVisibilitySetting: false,
-      showResponsesTo: [],
-      showGiverNameTo: [],
-      showRecipientNameTo: [],
-    });
+    // do not reset the visibility settings if reverting feedback path to preset template provided
+    if (this.model.isUsingOtherFeedbackPath && this.isCustomFeedbackVisibilitySettingAllowed &&
+        this.model.giverType === giverType && this.model.recipientType === newRecipientType) {
```
Comment (moziliar) [inline src/web/app/components/question-edit-form/question-edit-form.component.ts:251]: Can I check with you if all the conditions here are necessary? E.g. checking `this.isCustomFeedbackVisibilitySettingAllowed`

File: src/web/app/components/question-edit-form/question-edit-form.component.ts [hunk 2]
```diff
@@ -246,15 +246,25 @@ export class QuestionEditFormComponent implements OnInit {
...
+      this.triggerModelChangeBatch({
+        giverType,
+        isUsingOtherFeedbackPath: false,
+        recipientType: newRecipientType,
```
Comment (moziliar) [inline src/web/app/components/question-edit-form/question-edit-form.component.ts:255]: Do you need to trigger an update of `recipientType` if it already is `newRecipientType` by the condition above? Likewise for `giverType`.
The update here following the equality checks seems redundant.

**Review by moziliar — CHANGES_REQUESTED** [review]
Sorry for the late review. I roughly get the problem and your solution which fixes the issue of `isUsingOtherFeedbackPath` not being updated to true in this workflow. Can you help to address some of my doubts above?

**Comment by stapletonce** [comment]
@moziliar I agree, both of those areas seem redundant to me. I have updated the code and re-performed the tests with those changes.

**Comment by moziliar** [comment]
Can I check what would happen if the user clicks `custom feedback path` again in the dropdown when the ui is already in `custom feedback path`? Logically, this code looks to me that it would just toggle the template back to default?

**Comment by stapletonce** [comment]
https://user-images.githubusercontent.com/55817381/113047982-01f30300-9170-11eb-8305-798a079fd3cb.mov

It does not toggle back to default with these changes @moziliar

**Comment by moziliar** [comment]
Yup, tested locally and the change seems to solve the problem partially.

Can you help to look into this problem too?
Mar-31-2021 13-26-41

**Comment by stapletonce** [comment]
@moziliar What is the desired behavior here? I was under the impression that the visibility should only stay the same when switching between custom and template when the custom and template matched. In your example, you switch from Students in this course -> Giver (Self feedback) to Students in this course -> Giver's teammates so visibility resets.

**Comment by moziliar** [comment]
@stapletonce Sorry I didn't specify what I sense wrong: notice how the first time I click to Giver's teammates template for the first time, the custom template part still stays, whereas the second time triggers the revert to the template display. The behaviour seems correct for people who are involve in fixing it, but might be confusing to new instructor has not much visual change has been rendered for the first click (the second is a lot more drastic than the first).

Maybe we could just make them the same as discussed in the issue:

> User interface reverts to template display prior to selecting custom feedback path

**Comment by stapletonce** [comment]
@moziliar Ah, I see. Thank you for clearing this up - I can work on this now!

**Comment by moziliar** [comment]
@stapletonce hi, is this PR ready for review?

**Comment by stapletonce** [comment]
yes please!

File: src/web/app/components/question-edit-form/question-edit-form.component.ts [hunk 3]
```diff
@@ -246,15 +246,26 @@ export class QuestionEditFormComponent implements OnInit {
...
-      recipientType: newRecipientType,
-      commonVisibilitySettingName: 'Please select a visibility option',
-      isUsingOtherVisibilitySetting: false,
-      showResponsesTo: [],
-      showGiverNameTo: [],
-      showRecipientNameTo: [],
-    });
+    if (this.model.giverType === giverType && this.model.recipientType === newRecipientType) {
+      // do not reset the visibility settings if reverting feedback path to preset template provided
+      if (this.model.isUsingOtherFeedbackPath) {
```
Comment (moziliar) [inline src/web/app/components/question-edit-form/question-edit-form.component.ts:251]: It is perhaps not a good practice to have two such if's as they create an arrow-antipattern. You might want to just do 
```suggestion
    if (this.model.giverType === giverType && this.model.recipientType === newRecipientType
      && this.model.isUsingOtherFeedbackPath) {
```
in which the precedence of the booleans achieves the exact same effect.

**Review by moziliar — CHANGES_REQUESTED** [review]
One minor part. The rest look ok!

**Comment by moziliar** [comment]
Sorry I take that back... it seems like the logic is wrong in this case🥲 you can revert that and the rest LGTM

File: src/web/app/components/question-edit-form/question-edit-form.component.ts (hunk 3 above)
Comment (moziliar) [reply src/web/app/components/question-edit-form/question-edit-form.component.ts]: Wrong logic here. Shouldn't omit the path where the function does nothing

**Review by moziliar — APPROVED** [review]
LGTM. Wait for CI to complete

**Comment by stapletonce** [comment]
@moziliar success 🎉

**Comment by stapletonce** [comment]
@Derek-Hardy ready for review

**Review by Derek-Hardy — APPROVED** [review]
LGTM, thanks for the contribution.

**Review by madanalogy — APPROVED** [review]

**Pull request merged**